rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,492 @@
+    # ... implementation
+}
+```
+
+## Current problems
+
+### Inconsistent usage outside /lib folder
+
+Those comments are only used and parsed consistently in the /lib folder. Outside of this folder the format doesn't follow the convention strictly. Also the comments outside /lib are not used to generate any output.
+
+### Unspecified format
+
+In general, the format for writing documentation strings is **not specified**.
+
+The *nixdoc*-tool enforces a somewhat consistent format but the actual format was never specified and is not enforced in parts of nixpkgs where nixdoc is currently not applied.
+
+Extending the scope of *nixdoc* is not the primary goal. Instead, we should find formal rules for writing *doc-comments*. Tools like *nixdoc* can then implement against this RFC instead of the format relying on nixdoc implementation details.
+
+### Only specific placements work
+
+The placement of those comments requires precisely commenting at the attribute set containing the function declarations, which is not usable for general-purpose documentation strings.
+
+e.g.,
+
+- file that directly exports the lib-function without wrapping it in an attribute set.
+- file that exports a constant expression
+- files outside of lib cannot be rendered due to missing conventions
+
+### Differentiate from regular comments
+
+The format doesn't allow any distinction between doc-comments or regular comments.
+
+Having a distinction would allow us to
+
+1. Find all comments that are part of the documentation
+2. Render them in the documentation format
+3. Connect the documentation to the exact places in the nix code. This is already done, but only for nixpkgs/lib.
+
+### References to the problems above
+
+#### nixpkgs - Dosctrings examples
+
+- [lib/attrsets](https://github.com/NixOS/nixpkgs/blob/master/lib/attrsets.nix)
+- [trivial-builders](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/trivial-builders.nix)
+- [stdenv/mkDerivation](https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/make-derivation.nix)
+- [nixos/lib/make-disk-image](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix)
+- more...
+
+#### frameworks
+
+- [dream2nix/utils](https://github.com/nix-community/dream2nix/blob/main/src/utils/config.nix)
+- [dream2nix/templates/builder](https://github.com/nix-community/dream2nix/blob/main/src/templates/builders/default.nix)
+- more... 
+
+#### Other tools
+
+Other tools that work directly with the nix AST and comments:
+
+- [noogle](https://noogle.dev) - Nix API search engine. It allows to search functions and other expressions.
+- [nix-doc](https://github.com/lf-/nix-doc) - A Nix developer tool leveraging the rnix Nix parser for intelligent documentation search and tags generation
+- [manix](https://github.com/mlvzk/manix) - A fast CLI documentation searcher for nix.
+
+# Detailed design
+[design]: #detailed-design
+
+**Proposed Solution:**
+
+Use `##` For doc-comment body and markdown within
```

Comment:
Being a bit pedantic, in Nix a Python-like comments can start in any column (going till the end of line).

Are you proposing any such comment as a possible doc-comment? Or only the ones that start in the first column?

```nix
let
   x = 1; ## this is not a doc-comment
   y = 2; ## this is not a doc-comment
in
## this is a doc-comment
  x+y
  ## this is not a doc-comment
```